core/storage/suggestion/gae_models.py

```diff
@@ -0,0 +1,292 @@
+# Copyright 2018 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for Oppia suggestions."""
+
+from core.platform import models
+import feconf
+
+from google.appengine.ext import ndb
+
+(base_models,) = models.Registry.import_models([models.NAMES.base_model])
+
+# Constants defining the types of suggestions.
+SUGGESTION_TYPE_ADD = 'add'
+SUGGESTION_TYPE_EDIT = 'edit'
+
+SUGGESTION_TYPE_CHOICES = [
+    SUGGESTION_TYPE_ADD,
+    SUGGESTION_TYPE_EDIT
+]
+
+# Constants defining types of entities to which suggestions can be created.
+ENTITY_TYPE_EXPLORATION = 'exploration'
+ENTITY_TYPE_QUESTION = 'question'
+ENTITY_TYPE_SKILL = 'skill'
+ENTITY_TYPE_TOPIC = 'topic'
+
+ENTITY_TYPE_CHOICES = [
+    ENTITY_TYPE_EXPLORATION,
+    ENTITY_TYPE_QUESTION,
+    ENTITY_TYPE_SKILL,
+    ENTITY_TYPE_TOPIC
+]
+
+# Constants defining the different possible statuses of a suggestion.
+STATUS_ACCEPTED = 'accepted'
+STATUS_IN_REVIEW = 'review'
+STATUS_INVALID = 'invalid'
+STATUS_REJECTED = 'rejected'
+
+STATUS_CHOICES = [
+    STATUS_ACCEPTED,
+    STATUS_IN_REVIEW,
+    STATUS_INVALID,
+    STATUS_REJECTED
+]
+
+# Constants defining suggestion sub-types for each of the above types.
+SUGGESTION_EDIT_STATE_CONTENT = 'edit_exploration_state_content'
+
+SUGGESTION_SUB_TYPES = {
+    SUGGESTION_TYPE_ADD: [],
+    SUGGESTION_TYPE_EDIT: [
+        SUGGESTION_EDIT_STATE_CONTENT
+    ]
+}
+
+SUGGESTION_SUB_TYPE_CHOICES = [
+    SUGGESTION_EDIT_STATE_CONTENT
+]
+
+# Defines what is the minimum role required to review suggestions
+# of a particular sub-type.
+SUGGESTION_MINIMUM_ROLE_FOR_REVIEW = {
+    SUGGESTION_EDIT_STATE_CONTENT: feconf.ROLE_ID_EXPLORATION_EDITOR
+}
+
+
+class SuggestionParams(ndb.Model):
+    """Model to store all the suggestion params for a suggestion."""
+
+    # The contribution type. Either translation related or content related.
+    contribution_type = ndb.StringProperty(required=True)
+    # The contribution domain. The language of the translation or the subject
+    # category of the content being suggested.
+    contribution_domain = ndb.StringProperty(required=True)
```

Comment:
For content related changes to an exploration I thought an apt parameter would be the subject category (like algebra, algorithms, architecture, etc). For translation changes the language would be the apt parameter. And for images I don't think there will be a necessity for a domain (because a graphic designer would not be bound by subject constraints). For adding questions to skills, I still need to decide on a parameter, but probably the best one there would be the topic the skill is under?

So in all these cases (which were the concrete cases we described), just a string would work fine for the domain, and one more string for the type of contribution - translation, graphic, content, question would be all the possible choices (I think) for the type.

Reply:
I'm a little confused about what the parameter is used for. In the first case, for example, why do you need the subject category? How will it be used?

Reply:
It will be used to know what category the suggester will be "scored" upon.

Reply:
Ah, I see.

Are all "contribution domains" interpreted similarly (i.e., as a key for "score")? If so, this should be made clear in the docstring.

Reply:
Yes. I'll edit the docstring appropriately. The scoring key will be a combination of both contribution_type and the contribution_domain. I'll indicate this in the docstring.

Reply:
Edited the docstring.